twenty-first/src/math/polynomial.rs

```diff
@@ -2583,4 +3084,396 @@ mod test_polynomials {
+    fn reverse_polynomial_with_zero_constant_term_twice_gives_shift_back(
+        #[filter(!#f.is_zero())] f: Polynomial<BFieldElement>,
+    ) {
+        let fx_plus_1 = f.shift_coefficients(1);
+        prop_assert_ne!(fx_plus_1.clone(), fx_plus_1.reverse().reverse());
+        prop_assert_eq!(
+            fx_plus_1.clone(),
+            fx_plus_1.reverse().reverse().shift_coefficients(1)
+        );
+    }
+
+    #[proptest]
+    fn reduce_by_structured_modulus_and_reduce_long_division_agree(
+        #[strategy(1usize..10)] n: usize,
+        #[strategy(1usize..10)] m: usize,
+        #[strategy(vec(arb(), #m))] b_coefficients: Vec<BFieldElement>,
+        #[strategy(1usize..100)] _deg_a: usize,
+        #[strategy(vec(arb(), #_deg_a + 1))] _a_coefficients: Vec<BFieldElement>,
+        #[strategy(Just(Polynomial::new(#_a_coefficients)))] a: Polynomial<BFieldElement>,
+    ) {
+        let mut full_modulus_coefficients = b_coefficients.clone();
+        full_modulus_coefficients.resize(m + n + 1, BFieldElement::from(0));
+        *full_modulus_coefficients.last_mut().unwrap() = BFieldElement::from(1);
+        let full_modulus = Polynomial::new(full_modulus_coefficients);
+
+        let long_remainder = a.reduce_long_division(&full_modulus);
+        let structured_remainder = a.reduce_by_structured_modulus(&full_modulus);
+
+        prop_assert_eq!(long_remainder, structured_remainder);
+    }
+
+    #[test]
+    fn reduce_by_structured_modulus_and_reduce_agree_long_division_concrete() {
+        let a = Polynomial::new(
+            [1, 0, 0, 3, 4, 3, 1, 5, 1, 0, 1, 2, 9, 2, 0, 3, 1]
+                .into_iter()
+                .map(BFieldElement::new)
+                .collect_vec(),
+        );
+        let mut full_modulus_coefficients =
+            [5, 6, 3].into_iter().map(BFieldElement::new).collect_vec();
+        let m = full_modulus_coefficients.len();
+        let n = 2;
+        full_modulus_coefficients.resize(m + n + 1, BFieldElement::from(0));
+        *full_modulus_coefficients.last_mut().unwrap() = BFieldElement::from(1);
+        let full_modulus = Polynomial::new(full_modulus_coefficients);
+
+        let long_remainder = a.reduce_long_division(&full_modulus);
+        let structured_remainder = a.reduce_by_structured_modulus(&full_modulus);
+
+        assert_eq!(
+            long_remainder, structured_remainder,
+            "naive: {}\nstructured: {}",
+            long_remainder, structured_remainder
+        );
+    }
+
+    #[proptest]
+    fn reduce_by_ntt_friendly_modulus_and_reduce_long_division_agree(
+        #[strategy(1usize..10)] m: usize,
+        #[strategy(vec(arb(), #m))] b_coefficients: Vec<BFieldElement>,
+        #[strategy(1usize..100)] _deg_a: usize,
+        #[strategy(vec(arb(), #_deg_a + 1))] _a_coefficients: Vec<BFieldElement>,
+        #[strategy(Just(Polynomial::new(#_a_coefficients)))] a: Polynomial<BFieldElement>,
+    ) {
+        if !b_coefficients.iter().all(|c| c.is_zero()) {
```

Comment:
This is confusing me. Is this checking whether polynomial `b` would be zero? If so, how about
```suggestion
		let b = Polynomial::new(b_coefficients);
		if b.is_zero() {            
			return Err(TestCaseError::Reject("some reason".into()));
		}
```
or an equivalent `#[filter(…)]`?